src/Orleans/Messaging/Message.cs

```diff
@@ -881,5 +788,581 @@ internal void DropExpiredMessage(MessagingStatisticsGroup.Phase phase)
+                NONE = 0,
+                ALWAYS_INTERLEAVE = 1 << 0,
+                CACHE_INVALIDATION_HEADER = 1 << 1,
+                CATEGORY = 1 << 2,
+                CORRELATION_ID = 1 << 3,
+                DEBUG_CONTEXT = 1 << 4,
+                DIRECTION = 1 << 5,
+                EXPIRATION = 1 << 6,
+                FORWARD_COUNT = 1 << 7,
+                NEW_GRAIN_TYPE = 1 << 8,
+                GENERIC_GRAIN_TYPE = 1 << 9,
+                RESULT = 1 << 10,
+                REJECTION_INFO = 1 << 11,
+                REJECTION_TYPE = 1 << 12,
+                READ_ONLY = 1 << 13,
+                RESEND_COUNT = 1 << 14,
+                SENDING_ACTIVATION = 1 << 15,
+                SENDING_GRAIN = 1 <<16,
+                SENDING_SILO = 1 << 17,
+                IS_NEW_PLACEMENT = 1 << 18,
+
+                TARGET_ACTIVATION = 1 << 19,
+                TARGET_GRAIN = 1 << 20,
+                TARGET_SILO = 1 << 21,
+                TARGET_OBSERVER = 1 << 22,
+                IS_UNORDERED = 1 << 23,
+                REQUEST_CONTEXT = 1 << 24,
+                // Do not add over int.MaxValue of these.
+            }
+
+            private Header headers = default(Header);
+
+            private Categories _category;
+            private Directions? _direction;
+            private bool _isReadOnly;
+            private bool _isAlwaysInterleave;
+            private bool _isUnordered;
+            private CorrelationId _id;
+            private int _resendCount;
+            private int _forwardCount;
+            private SiloAddress _targetSilo;
+            private GrainId _targetGrain;
+            private ActivationId _targetActivation;
+            private GuidId _targetObserverId;
+            private SiloAddress _sendingSilo;
+            private GrainId _sendingGrain;
+            private ActivationId _sendingActivation;
+            private bool _isNewPlacement;
+            private ResponseTypes _result;
+            private DateTime? _expiration;
+            private string _debugContext;
+            private List<ActivationAddress> _cacheInvalidationHeader;
+            private string _newGrainType;
+            private string _genericGrainType;
+            private RejectionTypes _rejectionType;
+            private string _rejectionInfo;
+            private Dictionary<string, object> _requestContextData;
+
+            public Header Headers => headers;
+
+            public Categories Category
+            {
+                get { return _category; }
+                set
+                {
+                    _category = value;
+                    headers = headers | Header.CATEGORY;
+                }
+            }
+
+            public Directions? Direction
+            {
+                get { return _direction; }
+                set
+                {
+                    _direction = value;
+                    headers = value == null ? headers & ~Header.DIRECTION : headers | Header.DIRECTION;
```

Comment:
suggest adding value == default(x) checks to every setter to set/clear flags.  Seem to be doing is selectively.